### PR TITLE
Binary Insertion Sort

### PR DESCRIPTION
**Pull request by syedjafer**
### Describe your change:
Implementation of binary insertion sort. Binary insertion sort is a sorting algorithm similar to insertion sort, but instead of using linear search to find the position where the element should be inserted, we use binary search. Thus, we reduce the number of comparisons for inserting one element from O(N) (Time complexity in Insertion Sort) to O(log N).

* [X] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [X] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [X] This pull request is all my own work -- I have not plagiarized.
* [X] I know that pull requests will not be merged if they fail the automated tests.
* [X] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [X] All new Python files are placed inside an existing directory.
* [X] All filenames are in all lowercase characters with no spaces or dashes.
* [X] All functions and variable names follow Python naming conventions.
* [X] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [X] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [X] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [X] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$6537}`.

### CODE REVIEW

File: sorts/binary_insertion_sort.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    mid = (start + end) // 2
+    if array[mid] < key:
+        return binary_search(array, key, mid + 1, end)
+    elif array[mid] > key:
+        return binary_search(array, key, start, mid - 1)
+    else:
+        return mid
+
+
+def binary_insertion_sort(array: list[int]):
```
[the remark this answers is not in the document]
Comment (syedjafer) [reply sorts/binary_insertion_sort.py]: added return types

**Comment by sofezticated-gentleman** [comment]
@syedjafer i think so you dont need to add __name__

**Comment by syedjafer** [comment]
> @syedjafer i think so you dont need to add **name**

Can u pls explain this ?

**Review by codered1234 — APPROVED** [review]

**Comment by syedjafer** [comment]
@sofezticated-gentleman  Can you please review this.

**Review by CaedenPH — COMMENTED** [review]
@syedjafer I like the idea of this algorithm because it uses a different approach (recursion) to the current `binary_insertion_sort` (iterative) (https://github.com/TheAlgorithms/Python/blob/master/sorts/binary_insertion_sort.py) 

Would it be possible for you to add this implementation to the existing implementation instead of creating a new file for it?

@tianyizheng02 What is your opinion?

**Review by tianyizheng02 — COMMENTED** [review]
@CaedenPH I'm hesitant to add this implementation for the following reasons:

1) While this implementation and the existing implementation implement binary search differently, the most salient part of the algorithm (the sorting) is still basically identical. If we were to add this implementation to the existing file, it'd only really consist of an alternate binary search function. In any case, we already have both iterative and recursive implementations of binary search in this repo.

2) This implementation is suspiciously similar to the first Python implementation in this [GeeksForGeeks article](https://www.geeksforgeeks.org/binary-insertion-sort/). The variable names are different, but the structure is completely identical. While it _could_ be a coincidence, it's similar enough to be suspicious IMO.

**Comment by tianyizheng02** [comment]
Closing this PR due to inactivity. @syedjafer, like @CaedenPH said, even if we were to accept your implementation, it should be added alongside the existing implementation rather than overwriting it.

**Pull request closed without merging**